text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,655 @@
+workspace of a lot of crates makes their own homebrew script for updating
+versions and managing updates/publications. It'd be quite convenient if we could
+standardize across the Rust ecosystem how to manage this information!
+
+## Duplication in crate metadata
+
+The last primary area of duplication that this proposal attempts to tackle is in
+crate metadata in the `[package]` section. This includes items such as:
+
+```toml
+[package]
+authors = []
+license = "..."
+repository = "..."
+documentation = "..."
+```
+
+These metadata directives are often duplicated amongst all crates, especially
+author/license/repository information. This is a pretty poor experience if you'e
+got to keep writing down the information in so many places!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Cargo's manifest parsing will be updated with new features to support
+deduplicating each of the areas above. While all of these new features are
+pretty small in their own right, they all add up to greatly reducing the
+overhead of managing a workspace of many crates. The list of new features in
+Cargo will look like the following:
+
+## Workspace-level Dependencies
+
+The `[workspace]` section can now have a `dependencies` section which works the
+same way as the `[dependencies]` section in `Cargo.toml`:
+
+```toml
+# in workspace's Cargo.toml
+[workspace.dependencies]
+log = "0.3.1"
+log2 = { version = "2.0.0", package = "log" }
+serde = { git = 'https://github.com/serde-rs/serde" }
+wasm-bindgen-cli = { path = "crates/cli" }
+```
+
+Each workspace member can then reference this section in the workspace with a
+new dependency directive:
+
+```toml
+# in a workspace member's Cargo.toml
+[dependencies]
+log = { workspace = true }
+```
+
+This directive indicates that the `log` dependency should be looked up from
+`workspace.dependencies` in the workspace root. You can reference any name
+defined in `[workspace.dependencies]` too:
+
+```toml
+[dependencies]
+log2 = { workspace = true }
+```
+
+## No longer need both `version` and `path` to publish to Crates.io
+
+When you have a `path` dependency, Cargo's current behavior on publication looks
+like this:
+
+* If you have a `version` specifier as well, then the `path` key is deleted and
+  the crate is uploaded with the specified `version` as a dependency
+  requirement.
+* If you don't have a `version` specifier, then the dependency directive is
+  deleted and crates.io will not learn about this dependency. This is only
+  really useful for `dev-dependencies`.
+
+Cargo's behavior will change in this second case, instead following new logic
+for a missing `version` specifier. For dev-dependencies where the referenced
+package is `publish = false`, then the dependency will be dropped. Otherwise
+Cargo will assume that `version = "$dependency_version"` was specified, meaning
+that it requires at least the current version and otherwise any
+semver-compatible version.
```

Comment:
I don't think this is a good behaviour. I will try my best to illustrate the problem in it with an example below.

We have the following crates in our clap's workspace: `clap` and `clap_derive`

`clap` depends on `clap_derive`. `clap`'s `Cargo.toml` should look like the following:

```toml
[package]
version = "3.0.0"

[dependencies]
clap_derive = { path = "./clap_derive", version = "3.0.0" }
```

If we adopt this behaviour, we will be removing the `version` specifier.

```toml
[package]
version = "3.0.0"

[dependencies]
clap_derive = { path = "./clap_derive" }
```

Now, imagine that we fix a small bug in `clap` and increase the patch version and want to release it.

```toml
[package]
version = "3.0.1"

[dependencies]
clap_derive = { path = "./clap_derive" }
```

What this proposal means is that `clap@3.0.1` depends on `clap_derive@3.0.1` which does not exist. This would make us publish `clap_derive@3.0.1` which does not have any changes compared with `clap_derive@3.0.0`. This is just increasing the noise and resource wastage which IMHO is not good.

I think there's value in keeping the `version` specifier as mandatory.

> [Lerna](https://lerna.js.org) which is a popular JS workspace manager solved this issue by updating the versions of intra-workspace dependencies only when there's a change detected. I copied the logic into [cargo-workspaces](https://github.com/pksunkara/cargo-workspaces).

Reply:
I think the intent is for `dependency_version` to be the version of the dependent crate. I.E. the version of `clap_derive` in the toml at path, not the version of `clap`. I can see why the text wuld need to be clerere.

Reply:
> What this proposal means is that clap@3.0.1 depends on clap_derive@3.0.1 which does not exist. This would make us publish clap_derive@3.0.1 which does not have any changes compared with clap_derive@3.0.0. 

This is exactly the process that I follow with my crates, so I think it is reasonable.

Reply:
> I think the intent is for dependency_version to be the version of the dependent crate.

In that case, I think @alexcrichton can explain a bit more.

> This is exactly the process that I follow with my crates, so I think it is reasonable.

As I said, unnecessary release might generate more noise and burden for end-users depending on their preferences. We should try to keep the releases to a minimum. And I did end up succeeding in making the above mentioned tool to do that.

Reply:
Ah yes to clarify @Eh2406 is right here, the synthesized version for `clap_derive` would be `version = "3.0.0"` since it uses the version of the `clap_derive` crate, not the outer crate.

You'll be able to bump `clap` without this inferred dependency changing.